### PR TITLE
Added ZippedImageDataset and MultiZippedImageDataset

### PR DESCRIPTION
**Pull request by d0i**
ZippedImageDataset to mitigate load caused by small file access. Especially on NFS environment it works far better than ordinary ImageDataset (4x faster in a simple benchmark in our environment).

unit test code is not yet written. Simple benchmark is in the following gist.
https://gist.github.com/d0i/d18518262069288f836a8cb2214a1f11

### CODE REVIEW

**Comment by okuta** [comment]
Please check [travis result](https://travis-ci.org/chainer/chainer/jobs/318546059).

File: chainer/datasets/image_dataset.py [hunk 1]
```diff
@@ -149,15 +156,86 @@ def get_example(self, i):
...
+    def get_example(self, i):
+        # we need to keep lock as small as possible
+        # in addition, PIL may seek() on the file -- zipfile won't support it
+
+        if self._zf_pid != os.getpid(): # XXX
+            self._zf_pid = os.getpid()
+            self._zf = zipfile.ZipFile(self._zipfilename)
+
+        
+        with self._zf.open(self._paths[i]) as fzobj:
```
Comment (okuta) [inline chainer/datasets/image_dataset.py]: How about `image_file_mem = self._zf.read(self._paths[i])` ?

File: chainer/datasets/image_dataset.py [hunk 2]
```diff
@@ -149,15 +156,86 @@ def get_example(self, i):
...
+    and other networked file systems. The user shall find good balance
+    between zipfile size and number of zipfiles (e.g. granularity)
+
+    Args:
+        zipfilenames (list of strings): List of zipped archive filename.
+        dtype: Data type of resulting image arrays.
+    """
+    def __init__(self, zipfilenames, dtype=numpy.float32):
+        self._zfs = [ZippedImageDataset(zipfilename, dtype) for zipfilename in zipfilenames]
+        self._zpaths_accumlens = []
```
Comment (okuta) [inline chainer/datasets/image_dataset.py]: How about following code?
```python 
       self._zpaths_accumlens = [0]
       zplen = 0
        for zf in self._zfs:
            zplen += len(zf)
            self._zpaths_accumlens.append(zplen)

    def get_example(self, i):
        tgt = bisect.bisect(self._zpaths_accumlens, i) - 1

        lidx = i - self._zpaths_accumlens[tgt]
        return self._zfs[tgt].get_example(lidx)
```

Comment (doipfn) [reply chainer/datasets/image_dataset.py]: agreed.

File: chainer/datasets/image_dataset.py (hunk 1 above)
Comment (doipfn) [reply chainer/datasets/image_dataset.py]: agreed. Thanks!

**Comment by okuta** [comment]
Thanks!
Please check travis result.

File: chainer/datasets/image_dataset.py [hunk 3]
```diff
@@ -23,6 +26,13 @@ def _read_image_as_array(path, dtype):
     return image
 
 
+def _postprocess_image(image):
+    if image.ndim == 2:
+        # image is greyscale
+        image = image[:, :, numpy.newaxis]
```
Comment (okuta) [inline chainer/datasets/image_dataset.py]: How about `return image[None, ...]` ?

**Comment by okuta** [comment]
jenkins, test this please.

File: chainer/datasets/image_dataset.py (hunk 3 above)
Comment (doipfn) [reply chainer/datasets/image_dataset.py]: I guess you mean `image[..., None]`?
I confirmed it's much faster. Thanks!

Comment (okuta) [reply chainer/datasets/image_dataset.py]: I proposed merge two operation, insert new axis and transpose.

File: chainer/datasets/image_dataset.py [hunk 4]
```diff
@@ -149,11 +156,87 @@ def get_example(self, i):
...
+        zipfilename (str): a string to point zipfile path
+        dtype: Data type of resulting image arrays
+    """
+
+    def __init__(self, zipfilename, dtype=numpy.float32):
+        self._zipfilename = zipfilename
+        self._zf = zipfile.ZipFile(zipfilename)
+        self._zf_pid = os.getpid()
+        self._dtype = dtype
+        paths = filter(lambda x: not x.endswith('/'), self._zf.namelist())
```
Comment (okuta) [inline chainer/datasets/image_dataset.py]: How about `self._path = [x for x in self._zf.namelist() if not x.endswith('/')]` ?

File: chainer/datasets/image_dataset.py [hunk 5]
```diff
@@ -149,11 +156,87 @@ def get_example(self, i):
...
+        self._paths = list(paths)
+
+    def __len__(self):
+        return len(self._paths)
+
+    def __reduce__(self):
```
Comment (okuta) [inline chainer/datasets/image_dataset.py]: Please remove this function.

**Comment by okuta** [comment]
I checked `zipfile` module. So, we have to use `lock` for multi-thread iterator.
Cloud you fix it?

**Comment by okuta** [comment]
flake8
```
.\chainer\datasets\image_dataset.py:253:1: W391 blank line at end of file
```

**Comment by okuta** [comment]
jenkins, test this please.

**Pull request merged**